File: leetcode/Dynamic_Programming/0062_Unique_Paths.py
class Solution(object):
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        # dp[i][j]: the number of paths from grid[i][j] to grid[m-1][n-1]
        dp = [[0] * n for _ in range(m)]
        for i in range(m):
            dp[i][n - 1] = 1
        for j in range(n):
            dp[m - 1][j] = 1

        for i in range(m - 2, -1, -1):
            for j in range(n - 2, -1, -1):
                dp[i][j] = dp[i + 1][j] + dp[i][j + 1]

        return dp[0][0]

File: leetcode/Dynamic_Programming/0062_Unique_Paths.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        if (m <= 0 || n <= 0)
            return 0;
        vector<vector<int>> pathCounts(m, vector<int>(n, 0));
        for (int i = 0; i < m; i++)
            pathCounts[i][n - 1] = 1;
        for (int i = 0; i < n; i++)
            pathCounts[m - 1][i] = 1;
        for (int i = m - 2; i >= 0; i--) {
            for (int j = n - 2; j >= 0; j--)
                pathCounts[i][j] = pathCounts[i + 1][j] + pathCounts[i][j + 1];
        }

        return pathCounts[0][0];
    }
};